### PR TITLE
Log assembly loads during task run

### LINKED ISSUE #6303: Consider logging assembly loads into the binlog
We often have issues when MSBuild binaries are loaded from the GAC etc. Or NuGet/SDK binaries loaded from the wrong location. I think it would be beneficial to subscribe to AppDomain.CurrentDomain.AssemblyLoad and log all assembly loads, including full path to the .dll, assembly name, MVID, and maybe some attributes such as AssemblyFileVersion (which often contains the Git SHA).

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #6303
Related: https://github.com/KirillOsenkov/MSBuildStructuredLog/pull/653

### Context

* Assembly loading logging for:
  * Task runs  - including loads in separate app domain for `AppDomainIsolatedTask`
  * Except for tasks runs in `OutOfProcessRaskHostNode` - as neither `LoggingService` nor `LoggingContext` are available there - **not sure if not supporting those is a problem**.
  * Sdk resolving (explicitly skipping resolvers defined in `Microsoft.Build` assembly)
  * Loggers initialization (explicitly skipping resolvers defined in `Microsoft.Build` assembly orwithin `Microsoft.Build.Logging` namespace) - **Note** - loggers initialiation reordered - Binlog is the first one - as assembly load events are emited at the time of each logger initialization, currently `LoggerService` dispatches events immediately after construction, so as a result order of logger definitions matter (only the earlier initialized gets messages about latter ones initialization assembly loads). Alternative would be to change the `LoggerService` contract to start dispatching on explicit request (easy to implement and preserving current MSBuild behavior - however might be potentially breaking for API users?) or to cache those events and log them later (doable, but convoluted - probably not worth the efforts now)
  * Evaluation

### Samples
* task run:

  ```
  Assembly loaded during TaskRun: System.Diagnostics.Debug, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: C:\Windows\Microsoft.Net\assembly\GAC_MSIL\System.Diagnostics.Debug\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Diagnostics.Debug.dll, MVID: bc6b825d-5f99-464e-a06d-e3ae4b860a34, AppDomain: [Default])
  ```
* evaluation:
  ```
  Assembly loaded during Evaluation: System.Collections.Immutable, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: C:\Program Files\dotnet\shared\Microsoft.NETCore.App\7.0.2\System.Collections.Immutable.dll, MVID: 5a4c54a3-2092-428e-89cc-a391fd9d398a, AppDomain: [Default])
  ```
* logger initialization:

  ```
  Assembly loaded during LoggerInitialization (MyLoggers.BasicFileLogger): System.Reflection.Metadata, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: C:\Program Files\dotnet\shared\Microsoft.NETCore.App\7.0.2\System.Reflection.Metadata.dll, MVID: c602a8ce-3890-40fc-aa9b-fa3c23f81aab, AppDomain: [Default])
  ```
* custom sdk resolver (no sample - similar to logger initialization)

### Perf impact

No observable impact.

 * `MSBUILDLOGALLASSEMBLYLOADS` is used to turn on loggin of all assembly loads
 * build command: `msbuild.exe <proj> /nodeReuse:false /bl`

| Scenario | Mean Duration |
| --- | --- |
| Orchard - MSBuild main | 00:00:59.25 |
| Orchard - curr branch | 00:00:58.30 |
| Orchard - curr branch, all loads  | 00:00:58.30 |
|  |  |
| Console - MSBuild main | 00:00:01.08 |
| Console - curr branch | 00:00:01.09 |
| Console - curr branch, all loads  | 00:00:01.10 |

### CODE REVIEW

File: src/Build/Resources/Strings.resx [hunk 1]
```diff
@@ -1942,9 +1942,9 @@ Utilization:          {0} Average Utilization: {1:###.0}</value>
     </comment>
   </data>
   <data name="ProjectCacheHitWithOutputs" xml:space="preserve">
-    <value>{0} -> Cache Hit</value>
```
Comment (JanKrivanek) [inline src/Build/Resources/Strings.resx:1945]: This is autogenerated - I'm wondering how invalid xml got here :thinking:

File: src/Build/BackEnd/Components/Logging/LoggingService.cs [hunk 2]
```diff
@@ -1594,6 +1595,8 @@ private void InitializeLogger(ILogger logger, IEventSource sourceForLogger)
         {
             try
             {
+                // TODO: can we distinguish builtin and custom loggers here?
```
Comment (JanKrivanek) [inline src/Build/BackEnd/Components/Logging/LoggingService.cs:1598]: Builtin are probably defined in assemblies that are anyway already loaded, so we might not care...
Any thoughts here are appreciated.

Same applies for SdkResolvers

**Comment by KirillOsenkov** [comment]
Don't forget to increment the file format version:
https://source.dot.net/#Microsoft.Build/Logging/BinaryLogger/BinaryLogger.cs,7791c3b6dbbeedbf,references

I'd say don't worry about the out of proc task node - it's a black box to us anyway, let's consider it a "tool" that we don't control.

Make sure AppDomain stuff works on Core (Core doesn't have AppDomains) or gracefully degrades.

I'm torn on whether we should include the built-in assembly loads. On one hand, we want to know where our own stuff is loaded from. On the other hand, we can already infer since we know the path to MSBuild.exe, and maybe not worth having the same redundant messages at the beginning of every binlog.

I'd say let's record a couple of medium-sized real-world binlogs when the PR is ready, add viewer support, look at the logs in the viewer and then tweak to achieve the desired behavior.

Also remember that we should submit the binlog viewer support for the new file format version first, test it with the test binlogs, I will publish the new viewer version, then we merge this PR. So that by the time early adopters start recording these binlogs the viewer already supports them.

See a sample PR by @edvilme that incremented the binlog version:
https://github.com/dotnet/msbuild/pull/8146

and the corresponding viewer PR:
https://github.com/KirillOsenkov/MSBuildStructuredLog/pull/644

**Comment by KirillOsenkov** [comment]
Forgot to say, the general direction looks good to me!

File: src/Build/Evaluation/Evaluator.cs [hunk 3]
```diff
@@ -1847,6 +1849,8 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
                 // Combine SDK path with the "project" relative path
                 try
                 {
+                    // TODO: can we distinguish builtin and custom sdkresolvers here?
+                    using var _ = AssemblyLoadsTracker.StartTracking(_evaluationLoggingContext, AssemblyLoadingContext.SdkResolution);
```
Comment (rainersigwald) [inline src/Build/Evaluation/Evaluator.cs:1853]: The only built-in one is `DefaultSdkResolver` so an explicit check for that should suffice.

File: src/Build/BackEnd/Components/Logging/LoggingService.cs (hunk 2 above)
Comment (rainersigwald) [reply src/Build/BackEnd/Components/Logging/LoggingService.cs:1598]: Feels like we _could_ have an explicit check for the known types, but there are enough of them that I don't know how helpful it would be.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            StopTracking();
+        }
+
+        private static IDisposable StartTracking(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:66]: Can you put this behind a 17.6 changewave opt-out, so if we discover some functional problem we can easily disable it?

File: src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs [hunk 5]
```diff
@@ -1101,6 +1101,11 @@ public override void MessageHandler(object sender, BuildMessageEventArgs e)
         {
             if (showOnlyErrors || showOnlyWarnings) return;
 
+            if (e.BuildEventContext == null && e is AssemblyLoadBuildEventArgs)
+            {
+                return;
```
Comment (rainersigwald) [inline src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs:1106]: Should we not log this in the text log, too? It's generally less helpful than binlogs but still has some place, and is (sadly) still popular.

File: src/Framework/AssemblyLoadBuildEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            LoadingContext = loadingContext;
+            AssemblyName = assemblyName;
+            AssemblyPath = assemblyPath;
+            MVID = mvid;
+            AppDomainId = appDomainId;
+            AppDomainFriendlyName = appDomainFriendlyName;
+        }
+
+        public AssemblyLoadingContext LoadingContext { get; private set; }
```
Comment (rainersigwald) [inline src/Framework/AssemblyLoadBuildEventArgs.cs:36]: This is a great representation for the task case, where the event context information is sufficient to understand what task did the load. In the other cases, it might be nice to have an association to "what plugin did the loading?". Only way I can see out of that is a string.

File: src/Framework/AssemblyLoadBuildEventArgs.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        public string AssemblyName { get; private set; }
+        public string AssemblyPath { get; private set; }
+        public Guid MVID { get; private set; }
+        public int AppDomainId { get; private set; }
+        public string AppDomainFriendlyName { get; private set; }
```
Comment (rainersigwald) [inline src/Framework/AssemblyLoadBuildEventArgs.cs:41]: Think it's worth anything to avoid storing and reporting these for the default app domain and/or on core where there are no app domains?

Would indicating `AssemblyLoadContext` on core be helpful at all (I don't think so at the moment)?

**Review by rainersigwald — COMMENTED** [review]
I think my big question is whether there's any noticeable perf cost. If there is, is it worth the additional information in general, or should we build an opt-in?

We shouldn't be loading assemblies all that much, so I'm hopeful we can just have it on . . .

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+//
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Build.BackEnd.Logging;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.BackEnd.Components.RequestBuilder
+{
+    internal class AssemblyLoadsTracker : MarshalByRefObject, IDisposable
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:12]:
```suggestion
    internal sealed class AssemblyLoadsTracker : MarshalByRefObject, IDisposable
```

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    {
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:13]: Does this type run on both core and framework? Heavy use of `AppDomain` makse me think framework but not seeing `#if` to remove it from core builds.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        private static IDisposable StartTracking(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            var tracker = new AssemblyLoadsTracker(loggingContext, loggingService, context, appDomain ?? AppDomain.CurrentDomain);
+            if (appDomain != null)
+            {
+                s_instances.AddOrUpdate(appDomain, tracker, (_, loadsTracker) => loadsTracker);
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:70]: The `AddOrUpdate` call seems suspicious here. In the case there is already an existing entry the value `tracker` is essentially lost. It will get a call to `StartTracking` but never `StopTracking`. I don't know the significance of that in the MSBuild system but it seems better to avoid if we can. 

It feels like the code here should be doing one of the following: 

1. Not calling `StartTracking` then the`AssemblyLoadsTracker` instance is not in the list
2. Having `s_instances` store a list vs. a single value

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            tracker.StartTracking();
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:72]: Note: consider the case where the same `AppDomain` is registered twice (the crux of my comment above). This code actually produces a subtle race condition. There is a period of time inside the call to `StartTracking` that assembly loads are not monitored and that means we could end up missing an entry in the log.

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs [hunk 12]
```diff
@@ -662,6 +663,7 @@ private void ExecuteIntrinsicTask(ItemBucket bucket)
         /// </summary>
         private async Task<WorkUnitResult> InitializeAndExecuteTask(TaskLoggingContext taskLoggingContext, ItemBucket bucket, IDictionary<string, string> taskIdentityParameters, TaskHost taskHost, TaskExecutionMode howToExecuteTask)
         {
+            using var assemblyLoadsTracker = AssemblyLoadsTracker.StartTracking(taskLoggingContext, AssemblyLoadingContext.TaskRun);
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:666]: Why not be specific about `AppDomain.CurrentDomain` here to make it explicit it's part of the tracking?

File: src/Framework/AssemblyLoadBuildEventArgs.cs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+#nullable disable
+
+using System;
+using System.IO;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Framework
+{
+    public class AssemblyLoadBuildEventArgs : BuildMessageEventArgs
```
Comment (jaredpar) [inline src/Framework/AssemblyLoadBuildEventArgs.cs:13]: Given this is a public type, is there a reason you all don't default to `sealed`?

File: src/Framework/AssemblyLoadingContext.cs [hunk 14]
```diff
@@ -0,0 +1,12 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Build.Framework;
```
Comment (jaredpar) [inline src/Framework/AssemblyLoadingContext.cs:4]: File scoped namespaces FTW 🚀

File: src/Framework/AssemblyLoadBuildEventArgs.cs (hunk 13 above)
Comment (rainersigwald) [reply src/Framework/AssemblyLoadBuildEventArgs.cs:13]: We haven't had this rule but we should.

File: src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs (hunk 5 above)
Comment (JanKrivanek) [reply src/Build/Logging/ParallelLogger/ParallelConsoleLogger.cs:1129]: **tl;dr;:** I believe it's out of scope currently.

It logs with exception of assembly loading for custom loggers. Those do not have any `BuildContext` that `ParallelConsoleLogger` relies upon quite a lot. It would take some refactoring of `ParallelConsoleLogger`, which is definitely doable but feels beyond scope of this change. 
If it would be missed - it can definitely be added later on.
Thoughts?

File: src/Framework/AssemblyLoadBuildEventArgs.cs (hunk 7 above)
Comment (JanKrivanek) [reply src/Framework/AssemblyLoadBuildEventArgs.cs:40]: Fair enough - updating to store/transfer null (single bit) for cases where `IsDefaultAppDomain()` returns `true`.
This handles core and framework and is prepared for cases like running task on core, but reporting the result to framework scheduler (not currently possible - but might be in the future) or vice versa; while using single bit of wire format for the most common cases.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs (hunk 9 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:12]: It's core and framework tested.
That being said - the dictionary with app domains is really needed just on FW - so putting this behing `#ifdef`
Rest I believe needs to stay as is - but it might be my lack of knowledge - for tracking assembly loads I need `AppDomain` (even if it's just the default one on core) to subscribe - correct?

**Comment by JanKrivanek** [comment]
Thank you for great suggestions!
I've tried to adress those (in https://github.com/dotnet/msbuild/pull/8316/commits/d08a0b4abaa1bff6732694fd282e27752657b19a), namely:

* Not observing loads triggered by builtin types (loggers, sdk resolvers)
* Putting the tracking and logging behinde ChangeWave
* Not storing/transfering/logging AppDomain details for default app domain (Core and Framework)
* Grace handle concurrent calls for multiple app domains
* Logging the type initiating the assembly loads (if known)
* Incrementing binloger version (I'm yet to create PR into the binlog viewer)

**Review by KirillOsenkov — APPROVED** [review]
I took a cursory look and everything I saw looks good. I didn't look too deeply though.

**Comment by JanKrivanek** [comment]
Coresponding Binlog viewer PR: https://github.com/KirillOsenkov/MSBuildStructuredLog/pull/653
@KirillOsenkov FYI

File: src/MSBuild/XMake.cs [hunk 15]
```diff
@@ -3210,8 +3210,12 @@ private static string ProcessToolsVersionSwitch(string[] parameters)
                 originalVerbosity = ProcessVerbositySwitch(verbositySwitchParameters[verbositySwitchParameters.Length - 1]);
                 verbosity = originalVerbosity;
             }
+            var loggers = new List<ILogger>();
+
+            var outVerbosity = verbosity;
+            ProcessBinaryLogger(binaryLoggerParameters, loggers, ref outVerbosity);
```
Comment (JanKrivanek) [inline src/MSBuild/XMake.cs:3216]: I reordered loggers initialization to have binlog on top - to ensure it receives all the events from initialiation of custom loggers.

@rainersigwald - is there any hidden risk associated to reordering the creation and initialiation of the loggers

**Comment by Forgind** [comment]
I'm wondering if we could simplify this. When do we expect to load assemblies? This PR seems to just say that _when_ we load assemblies, we should log it but does that with broad strokes. Could we have something that targets the places where we load an assembly and log that an assembly was loaded at those places? So for tasks, for instance, you could put things [here](https://github.com/dotnet/msbuild/blob/446f42e6d1478acb44faec787f90f57f38c08b73/src/Shared/TypeLoader.cs#L161) and [here](https://github.com/dotnet/msbuild/blob/446f42e6d1478acb44faec787f90f57f38c08b73/src/Shared/TypeLoader.cs#L189).

File: src/MSBuild/XMake.cs (hunk 15 above)
Comment (rainersigwald) [reply src/MSBuild/XMake.cs:3216]: Nothing springs to mind.

**Comment by rainersigwald** [comment]
@Forgind the .NET assembly loader(s) load assemblies just-in-time to provide definitions for methods that can be called in a method as it is JITted before its first run. Because of that, we cannot control when assemblies will be loaded beyond the plugin entrypoints we explicitly control.

I originally was thinking of logging only the resolved paths of plugin entrypoints, but this approach is more comprehensive and enables debugging collisions between transitive dependencies of plugins, so it seems superior to me.

**Comment by JanKrivanek** [comment]
This change is mostly about runtime induced (implicit) loading.
The explicit type loading is actualy not loaded.

For this reason (closer explained by @rainersigwald) we unfortunately need to mount observing for the entire execution time of the component in question.
For tasks and for evaluation this is fortunately possible to perform in properly constrained code section (and execution time).

For loggers and sdkresolvers, the typeloader should actually be the additional point in code hooked (I choosed not to hook, as for those two components we hopefully should know where we are loading from and this way I do not need to complicately distinguish which type loading is of interest and which to fileter out).

Additionaly for the loggers the code obvserves only the initialization section - in hopes all the assemblies will be touched there. But I can imagine esoteric loggers that will use additional assemblies only during handling of specific events or even shutdown - this is intentionally not supported (as it opens a can of worm of distniguishing what caused assembly load at specific point of time - e.g. when custom task is logging to custom logger).

**tl;dr;:** please let me know if you think the assembly loads caused by loading the actual types implementing the custom loggers or sdkresolvers should be logged as well (*in addition* to the current hooked usages of the components.)

**Comment by rainersigwald** [comment]
> **tl;dr;:** please let me know if you think the assembly loads caused by loading the actual types implementing the custom loggers or sdkresolvers should be logged as well (_in addition_ to the current hooked usages of the components.)

I think this is important. One way things can go wrong is a build that has references to two versions of a build task with the same identity. If they're loaded in separate worker nodes they can both work, but when loaded in the same process the first one "wins" the load and might cause behavior differences.

**Comment by JanKrivanek** [comment]
> I think this is important. One way things can go wrong is a build that has references to two versions of a build task with the same identity. If they're loaded in separate worker nodes they can both work, but when loaded in the same process the first one "wins" the load and might cause behavior differences.

The task case is hooked withing [`InitializeAndExecuteTask`](https://github.com/dotnet/msbuild/blob/main/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs#L657) - which is eventually calling into the `TypeLoader` - so the assembly information should be properly loged in this case. I'll verify with explicit custom task.

The `TypeLoader` loading is not hooked for logger and sdk resolver (the logging service is not yet available there) - though this can definitely be solved somehow (deffered log message?) if it feels as importnat usecase.


One another question that pops-up during testing this: can we somehow distinguish builtin tasks? (while possibly special case keep handling some that are used as surrogate for custom tasks - like `RoslynCodeTaskFactory`) Is there anything that can be used to filter those out?

image

**Comment by Forgind** [comment]
Can we just check which assembly it's from? If it's in a fixed list, then it's us. I think we have a "fixed" list somewhere in MSBuildLocator

**Comment by rainersigwald** [comment]
For tasks I think it's literally just `Microsoft.Build.Tasks.Core.dll`. Loggers we have in `Microsoft.Build.dll` though.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 16]
```diff
@@ -107,19 +83,47 @@ public void Dispose()
...
+        private static bool IsBuiltinType(string? typeName)
+        {
+            if (string.IsNullOrEmpty(typeName))
+            {
+                return false;
+            }
+
+            return typeName!.StartsWith("Microsoft.Build", StringComparison.OrdinalIgnoreCase) ||
+                   typeName.StartsWith("Microsoft.NET.Build", StringComparison.OrdinalIgnoreCase) ||
+                   typeName.StartsWith("Microsoft.CodeAnalysis.BuildTasks", StringComparison.OrdinalIgnoreCase);
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:95]: I don't think we should consider the Roslyn tasks built in. One of the biggest issues we have to investigate is why Roslyn tasks don't laod the way we expect precisely because they're not built in.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs [hunk 17]
```diff
@@ -107,19 +83,47 @@ public void Dispose()
...
         }
 
+        private static bool IsBuiltinType(string? typeName)
+        {
+            if (string.IsNullOrEmpty(typeName))
+            {
+                return false;
+            }
+
+            return typeName!.StartsWith("Microsoft.Build", StringComparison.OrdinalIgnoreCase) ||
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:93]: We should be using `StringComparison.Ordinal` here. Type names are case sensitive and known at compile time.

File: src/Build/BackEnd/Components/Logging/LoggingService.cs [hunk 18]
```diff
@@ -1589,8 +1591,17 @@ private void RouteBuildEvent(BuildEventArgs eventArg)
         /// <exception cref="Exception">Any exception which is a ExceptionHandling.IsCriticalException will not be wrapped</exception>
         private void InitializeLogger(ILogger logger, IEventSource sourceForLogger)
         {
+            ILogger UnwrapLoggerType(ILogger log)
+            {
+                return (log is ProjectCollection.ReusableLogger reusableLogger)
+                    ? reusableLogger.OriginalLogger
```
Comment (jaredpar) [inline src/Build/BackEnd/Components/Logging/LoggingService.cs:1597]: Is it possible for this type of nesting to be more than one level deep? Often for this type of problem there is an extension method defined called `Unwrap` that recursively calls itself to handle arbitrary nesting scenarios.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs (hunk 11 above)
Comment (jaredpar) [reply src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:136]: This convo was marked as resolved but I can't see how. The race condition still exists.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs (hunk 16 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs]: Makes sense - removed from filter.
Just to supplement the info - in case of need it's possible to force all tracking via defining `MSBUILDLOGALLASSEMBLYLOADS` env var.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs (hunk 17 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs]: This code can execute on top of the assembly filename as well (it's mounted prior actual loading of the task, when only the task name and some information about it's hosting assembly are known).
But since we are having consistent casing for our assembly files - it's good idea to be more strict here.

File: src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs (hunk 11 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs:136]: Thanks for reiterating this one!
Since the `StartTracking` instance method was private and was invoked only after construction - there was no chance for double subscription case - so the `unsubscribe-subscribe` was superfluous and I removed it. 
But it's a good point for future cases as well

File: src/Build/Evaluation/Evaluator.cs [hunk 19]
```diff
@@ -1840,9 +1843,16 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
                     }
                 }
 
+                IDisposable assemblyLoadTracker = null;
                 // Combine SDK path with the "project" relative path
                 try
                 {
+                    // Is the sdk resolver a custom type?
```
Comment (Forgind) [inline src/Build/Evaluation/Evaluator.cs:1850]: Aren't SDK resolvers in different assemblies? We have our built-in one, but NuGet owns theirs, C++ owns theirs, etc.

File: src/Build/Instance/TaskFactories/AssemblyTaskFactory.cs [hunk 20]
```diff
@@ -393,10 +400,14 @@ public void CleanupTask(ITask task)
...
 
 #if FEATURE_APPDOMAIN
-                if (taskAppDomain != null)
+                if (taskAppDomain != null && taskInstance != null)
                 {
                     _tasksAndAppDomains[taskInstance] = taskAppDomain;
                 }
+                else if (taskAppDomain != null)
+                {
+                    AssemblyLoadsTracker.StopTracking(taskAppDomain);
```
Comment (Forgind) [inline src/Build/Instance/TaskFactories/AssemblyTaskFactory.cs:409]: Do you still need _tasksAndAppDomains[taskInstance] = taskAppDomain; in this condition?

File: src/Shared/LogMessagePacketBase.cs [hunk 21]
```diff
@@ -139,7 +139,12 @@ internal enum LoggingEventType : int
         /// <summary>
         /// Event is a ResponseFileUsedEventArgs
         /// </summary>
-        ResponseFileUsedEvent = 20
+        ResponseFileUsedEvent = 20,
+
+        /// <summary>
+        /// Event is an AssemblyLoadBuildEventArgs
+        /// </summary>
+        AssemblyLoadEvent = 21,
```
Comment (Forgind) [inline src/Shared/LogMessagePacketBase.cs:147]: Have you tested if this works properly for a build that traverses multiple nodes? When introducing a new event args type, we typically have to implement (de)serialization in (Write/Read)Event(To/From)Stream. It's possible you can squeeze by because this is a message, but I'd want to make sure that's properly tested.

File: src/Build/Evaluation/Evaluator.cs (hunk 19 above)
Comment (JanKrivanek) [reply src/Build/Evaluation/Evaluator.cs:1850]: That's a good point - I believe that if it's not built-in, than the default tracking has a sense, as it might not be obvious from where it was loaded. (similarly to this comment about roslyn tasks: https://github.com/dotnet/msbuild/pull/8316#discussion_r1103814426). 
What do you think?

Btw. this code was accidentaly performin the filtering twice - I removed the superfluous call wrapping the `StartTracking`, as unified (and bit more comprehensive) filtering is performed within `StartTracking`.
It filters as well based on the typename - https://github.com/dotnet/msbuild/pull/8316/files/f6c696399c13667bc7742a8e94c0cfc3c79e46aa#diff-7f217b4ccb76592a2ab78dd7f8332509dfa2e2216d8a6032f956cbdf48df8d03R86-R96

File: src/Build/Instance/TaskFactories/AssemblyTaskFactory.cs (hunk 20 above)
Comment (JanKrivanek) [reply src/Build/Instance/TaskFactories/AssemblyTaskFactory.cs:409]: No - it's just a cleanup if the `taskInstance` is `null` and `taskAppDomain` is not (in such a case we do not want to track as task was not created).
The condition is altered as there was a hidden bug in the original code (indexing with `null` key `taskInstance`)

File: src/Shared/LogMessagePacketBase.cs (hunk 21 above)
Comment (Forgind) [reply src/Shared/LogMessagePacketBase.cs:147]: I normally think of reflection as a bit slow but sometimes required if we don't know anything about a build event. I'm wondering if there's some opportunity to optimize this a bit if we have some WriteToStream base method in the base type and have it take a BinaryWriter (and maybe a bool) and have it return a bool to indicate that we do or don't support self-serialization? I'm imagining that if it says yes, it would just call the method afterwards. Alternatively, it could just be a query to WriteToStream() and return true/false but not do any work. What do you think?

File: src/Build/Evaluation/Evaluator.cs (hunk 19 above)
Comment (JanKrivanek) [reply src/Build/Evaluation/Evaluator.cs]: There is a  opt-in trait (via `MSBUILDLOGALLASSEMBLYLOADS` env var) to stop any filtering.
So if for any reason any component of interest would not be hooked (e.g. due to being placed to 'Microsoft.Build' namespace), it can allways be opted-in.

File: src/Shared/LogMessagePacketBase.cs (hunk 21 above)
Comment (JanKrivanek) [reply src/Shared/LogMessagePacketBase.cs:147]: I totally agree the code would benefit from unifying (de)serialization techniques for cross-nodes communication - currently there are 3 different approaches: 
[MSBUILDLOGALLASSEMBLYLOADS ](https://github.com/dotnet/msbuild/blob/main/documentation/wiki/Nodes-Orchestration.md#messages-deserialization)

`ITranslatable` contract sounds as the best choice here - so ideally `BuildEventArgs` would implement and use that.

Though - that is totally out of scope of current change :-) Do you think any change to contract is needed in scope of this change? Do you want to crerate item for code/contracts refactoring?

I wouldn't worry about perf impact - reflection is definitely more costly then explicit contract implementation - though the cost is paid once per node, so I prefered to stick with current contracts

Comment (Forgind) [reply src/Shared/LogMessagePacketBase.cs:147]: It's once per node? I would've thought once per method per node. In any event, I can agree it's out of scope, and I can't think of anything else that is both clearly better and in scope. You can make the item or not—as you said, I don't think it'd be a high priority for us, so it might end up sitting for a while then getting closed as low priority. I think this otherwise looks good to me 🙂

**Review by Forgind — APPROVED** [review]

**Comment by KirillOsenkov** [comment]
I've published the viewer version that supports this, would be nice to test that it's able to open binlogs produced with this change.

**Comment by JanKrivanek** [comment]
@KirillOsenkov I confirm I can open my test logs (new console, orchard, etc.) with the latest official version :thumbsup:

I have an issue with some logs from previous version - I'm on it now.

Btw. a dumb question - Were there any thoughts to make the auto-update experience more friendly or understandable? I observed that the first open attempt led to error due to an old version, without simple info about update option/process. The Help->About menu presented a clear message about a new version and need to restart the app. Can this be perceived as confusing?

**Comment by KirillOsenkov** [comment]
Yeah we can certainly improve this. One of these days!

**Comment by JanKrivanek** [comment]
> Yeah we can certainly improve this. One of these days!

Since I complained :-) - let me try to propose a fix ...
